File: spoj/road.py
'''input
1
9 6
7 8 6
6 9 6
2 5 8
3 6 10
1 7 6
2 9 1
'''
def min(a,b):
	if a != 0 and a <= b:
		return a
	return b

tc = int(raw_input())

for t in xrange(tc):
	n,k = map(int,raw_input().split())

	ip = [""] * (k+1)
	for i in xrange(1,k+1):
		ip[i] = map(int, raw_input().split())

	# dp = [[0 for j in xrange(n+1)] for i in xrange(k+1)]
	# dp[0][0] = 0


	# for j in xrange(1,n+1):
	# 	for i in xrange(1,k+1):
	# 		if ip[i][0] < j and j <= ip[i][1]:
	# 			dp[i][j] = min(dp[i-1][j],dp[k][ip[i][0]] + ip[i][2])
	# 		else:
	# 			dp[i][j] = dp[i-1][j]

	# 	if not dp[k][j]:
	# 		break
	# if dp[k][n]:
	# 	print dp[k][n]
	# else:
	# 	print -1
	print n,k,ip
	ddp = [111 for i in  xrange(n+1)]
	ddp[0] = 0
	for i in xrange(1,k+1):
		start = ip[i][0]
		end = ip[i][1]
		cost = ip[i][2]
		for j in xrange(start,end+1):
			ddp[end] = min(ddp[end], ddp[j] + cost)
			print start,j,end,cost,ddp
	print ddp[n]

File: spoj/road.cpp
/*input
3
9 6
2 7 10
7 8 3
3 4 8
0 8 2
5 6 4
3 7 1

9 6
7 8 6
6 9 6
2 5 8
3 6 10
0 7 6
2 9 1

3 4
0 3 8
0 2 4
0 1 10
2 3 8
*/

#include <bits/stdc++.h>
using namespace std;

struct node {
    int start,end,cost;
};

const int inf=0x3f3f3f3f;
const int N=1010;

int dp[N];
node a[N];

bool comp(node a,node b) {
    return a.start < b.start;
}
int main() {
    int tt;
    scanf("%d",&tt);
    while(tt--) {
        int L,n;
        scanf("%d%d",&L,&n);
        for(int i=0;i<n;i++) {
            scanf("%d%d%d",&a[i].start,&a[i].end,&a[i].cost);
        }
        sort(a,a+n,comp);
        for(int i=0;i<=L;i++) {
            dp[i]=inf;
        }
        dp[0]=0;
        for(int i=0;i<n;i++) {
            for(int j=a[i].start;j<=a[i].end;j++) {
                dp[a[i].end] = min(dp[a[i].end] , dp[j]+a[i].cost);
            }
        }
        if(dp[L]==inf) {
            dp[L]=-1;
        }
        printf("%d\012",dp[L]);
    }
    return 0;
}
